pkg/authorization/admission/restrictusers/restrictusers.go

```diff
@@ -0,0 +1,279 @@
+func NewRestrictUsersAdmission(kclient kclientset.Interface) (admission.Interface, error) {
+	return &restrictUsersAdmission{
+		Handler: admission.NewHandler(admission.Create),
+		kclient: kclient,
+	}, nil
+}
+
+func (q *restrictUsersAdmission) SetOpenshiftClient(c oclient.Interface) {
+	q.oclient = c
+}
+
+// Admit makes admission decisions that enforce restrictions on adding
+// project-scoped role-bindings.  In order for a role binding to be permitted,
+// the project must be annotated to indicate that the subject of the role
+// binding is a collaborator on the project.
+func (q *restrictUsersAdmission) Admit(a admission.Attributes) (err error) {
+	if a.GetResource().GroupResource() != authorizationapi.Resource("rolebindings") {
+		return nil
+	}
+
+	// Ignore all operations that correspond to subresource actions.
+	if len(a.GetSubresource()) != 0 {
+		return nil
+	}
+
+	ns := a.GetNamespace()
+
+	// Ignore cluster-level resources.
+	if len(ns) == 0 {
+		return nil
+	}
+
+	binding, ok := a.GetObject().(*authorizationapi.RoleBinding)
+	if !ok {
+		return admission.NewForbidden(a,
+			fmt.Errorf("wrong object type: %t", a.GetObject()))
+	}
+
+	if !q.waitForSyncedStores(time.After(timeToWaitForCacheSync)) {
+		return admission.NewForbidden(a, errors.New("caches not synchronized"))
+	}
+
+	glog.V(4).Infof("Handling rolebinding %s/%s", binding.Namespace, binding.Name)
+
+	// Only bother checking deltas.
+	newSubjects := []kapi.ObjectReference{}
+	policy := authorizationapi.GetPolicyBindingName(binding.RoleRef.Namespace)
+	policybinding, _ := q.policyBindingLister.PolicyBindings(ns).Get(policy)
+	if policybinding != nil {
+		existingBinding := policybinding.RoleBindings[binding.Name]
+		if existingBinding != nil {
+			for _, newSubject := range binding.Subjects {
+				foundExisting := false
+				for _, existingSubject := range existingBinding.Subjects {
+					if existingSubject.Kind == newSubject.Kind &&
+						existingSubject.Name == newSubject.Name &&
+						existingSubject.Namespace == newSubject.Namespace {
+						foundExisting = true
+						break
+					}
+				}
+				if !foundExisting {
+					newSubjects = append(newSubjects, newSubject)
+				}
+			}
+		} else {
+			newSubjects = binding.Subjects
+		}
+	} else {
+		return admission.NewForbidden(a,
+			fmt.Errorf("no policy binding found: %s", policy))
+	}
+
+	if len(newSubjects) == 0 {
+		return nil
+	}
+
+	matchersList, err := q.oclient.CollaboratorMatchers(ns).List(kapi.ListOptions{})
+	if err != nil {
+		return admission.NewForbidden(a, err)
+	}
+
+	userLiterals := []string{}
+	userInGroupLiterals := []string{}
+	userSelectors := []labels.Selector{}
+	groupLiterals := []string{}
+	groupSelectors := []labels.Selector{}
+	serviceaccountLiterals := []string{}
+	serviceaccountInTheseNamespacesLiterals := []string{}
+
+	for _, matcher := range matchersList.Items {
+		// TODO: Use a cache for users, groups, and serviceaccounts.
+		switch {
+		case matcher.Spec.UserMatcher != nil:
```

Comment:
I'd like to see an interface defined that checks a subject.  Something like `Allowed(subject) error`.  I would be able to take a `collaboratormatcherspec` and get an allowed back out and then something that provides a union of `subjectcheckers`.  It'll make it easier to read.